File: lib/Render.js
/***************************************************************************************************************************************************************
 *
 * cfonts
 *
 * Sexy fonts for the console. (CLI output)
 *
 * @license     https://github.com/dominikwilkowski/cfonts/blob/master/LICENSE  GNU GPLv2
 * @author      Dominik Wilkowski  hi@dominik-wilkowski.com
 * @repository  https://github.com/dominikwilkowski/cfonts
 *
 * Render
 *   Main method to get the ANSI output for a string
 *
 **************************************************************************************************************************************************************/
'use strict';

var _require = require('./AddLetterSpacing.js'),
    AddLetterSpacing = _require.AddLetterSpacing;

var _require2 = require('./RenderConsole.js'),
    RenderConsole = _require2.RenderConsole;

var _require3 = require('./Debugging.js'),
    Debugging = _require3.Debugging,
    DEBUG = _require3.DEBUG;

var _require4 = require('./CharLength.js'),
    CharLength = _require4.CharLength;

var _require5 = require('./CheckInput.js'),
    CheckInput = _require5.CheckInput;

var _require6 = require('./CleanInput.js'),
    CleanInput = _require6.CleanInput;

var _require7 = require('./GetOptions.js'),
    GetOptions = _require7.GetOptions;

var _require8 = require('./AlignText.js'),
    AlignText = _require8.AlignText;

var _require9 = require('./Colorize.js'),
    Colorize = _require9.Colorize;

var _require10 = require('./AddLine.js'),
    AddLine = _require10.AddLine;

var _require11 = require('./AddChar.js'),
    AddChar = _require11.AddChar;

var _require12 = require('./GetFont.js'),
    GetFont = _require12.GetFont;

var _require13 = require('./constants.js'),
    CHARS = _require13.CHARS;

var _require14 = require('./Color.js'),
    Color = _require14.Color;

var _require15 = require('./Size.js'),
    Size = _require15.Size;

var _require16 = require('./Log.js'),
    Log = _require16.Log;
/**
 * Main method to get the ANSI output for a string
 *
 * @param  {string}  input       - The string you want to write out
 * @param  {object}  SETTINGS    - Settings object
 * @param  {boolean} debug       - A flag to enable debug mode
 * @param  {integer} debuglevel  - The debug level we want to show
 * @param  {object}  size        - The size of the terminal as an object, default: Size
 * @param  {integer} size.width  - The width of the terminal
 * @param  {integer} size.height - The height of the terminal
 *
 * @typedef  {object} ReturnObject
 *   @property {string}  string  - The pure string for output with all line breaks
 *   @property {array}   array   - Each line of output in an array
 *   @property {integer} lines   - The number of lines
 *   @property {object}  options - All options used
 *
 * @return {ReturnObject}        - CLI output of INPUT to be consoled out
 */


var Render = function Render(input) {
  var SETTINGS = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var debug = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : DEBUG.enabled;
  var debuglevel = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : DEBUG.level;
  var size = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : Size;
  Debugging.report("Running render", 1);
  DEBUG.enabled = debug;
  DEBUG.level = debuglevel;
  var INPUT = CleanInput(input, CHARS);
  var OPTIONS = GetOptions(SETTINGS);
  var output = []; // for output where each line is an output line

  var lines = 0; // for counting each line

  var FONTFACE = {}; // scoping the fontface object higher for fonts with just one color

  var _isGoodHuman = CheckInput(INPUT, OPTIONS.font, OPTIONS.colors, OPTIONS.background, OPTIONS.align, OPTIONS.gradient);

  if (!_isGoodHuman.pass) {
    Log.error(_isGoodHuman.message);
    return false;
  } // display an overview of options if debug flag is enabled


  if (DEBUG.enabled) {
    var outOption = "OPTIONS:\n  Text: ".concat(INPUT);

    for (var key in OPTIONS) {
      outOption += "\n  Options.".concat(key, ": ").concat(OPTIONS[key]);
    }

    Debugging.report(outOption, 3);
  }

  if (OPTIONS.font === 'console') {
    // console fontface is pretty easy to process
    FONTFACE = {
      colors: 1,
      lines: 1
    };
    var consoleOutput = RenderConsole(INPUT, OPTIONS, size);
    output = consoleOutput.output;
    lines = consoleOutput.lines;
  } else {
    // all other fontfaces need the font-file and some more work
    FONTFACE = GetFont(OPTIONS.font);

    if (!FONTFACE) {
      Log.error("Font file for the font \"".concat(OPTIONS.font, "\" could not be found.\nTry reinstalling this package."));
      return false;
    } // setting the letterspacing preference from font face if there is no user overwrite


    if (SETTINGS.letterSpacing === undefined) {
      Debugging.report("Looking up letter spacing from font face", 1);
      var width = 0;
      FONTFACE.letterspace.forEach(function (item) {
        var _char = item.replace(/(<([^>]+)>)/ig, ''); // get character and strip color infos


        if (width < _char.length) {
          width = _char.length;
        }
      });
      Debugging.report("Letter spacing set to font face default: \"".concat(width, "\""), 2);
      OPTIONS.letterSpacing = width;
    }

    var lineLength = CharLength(FONTFACE.buffer, FONTFACE.lines, OPTIONS); // count each output character per line and start with the buffer

    var maxChars = 0; // count each character we print for maxLength option

    output = AddLine([], FONTFACE.lines, FONTFACE.buffer, OPTIONS.lineHeight); // create first lines with buffer

    lines++;
    output = AddLetterSpacing(output, FONTFACE.lines, FONTFACE.letterspace, FONTFACE.colors, OPTIONS.colors, OPTIONS.letterSpacing); // add letter spacing to the beginning

    lineLength += CharLength(FONTFACE.letterspace, FONTFACE.lines, OPTIONS) * OPTIONS.letterSpacing; // count the space for the letter spacing

    for (var i = 0; i < INPUT.length; i++) {
      // iterate through the message
      var CHAR = INPUT.charAt(i).toUpperCase(); // the current character we convert, only upper case is supported at this time

      var lastLineLength = lineLength; // we need the lineLength for alignment before we look up if the next char fits

      Debugging.report("Character found in font: \"".concat(CHAR, "\""), 2);

      if (CHAR !== "|") {
        // what will the line length be if we add the next char?
        lineLength += CharLength(FONTFACE.chars[CHAR], FONTFACE.lines, OPTIONS); // get the length of this character

        lineLength += CharLength(FONTFACE.letterspace, FONTFACE.lines, OPTIONS) * OPTIONS.letterSpacing; // new line, new line length
      } // jump to next line after OPTIONS.maxLength characters or when line break is found or the console windows would have ran out of space


      if (maxChars >= OPTIONS.maxLength && OPTIONS.maxLength != 0 || CHAR === "|" || lineLength > size.width) {
        lines++;
        Debugging.report("NEWLINE: maxChars: ".concat(maxChars, ", ") + "OPTIONS.maxLength: ".concat(OPTIONS.maxLength, ", ") + "CHAR: ".concat(CHAR, ", ") + "lineLength: ".concat(lineLength, ", ") + "Size.width: ".concat(size.width, " "), 2);
        output = AlignText(output, lastLineLength, FONTFACE.lines, OPTIONS.align, size); // calculate alignment based on lineLength

        lineLength = CharLength(FONTFACE.buffer, FONTFACE.lines, OPTIONS); // new line: new line length

        lineLength += CharLength(FONTFACE.letterspace, FONTFACE.lines, OPTIONS) * OPTIONS.letterSpacing; // each new line starts with letter spacing

        if (CHAR !== "|") {
          // if this is a character and not a line break
          lineLength += CharLength(FONTFACE.chars[CHAR], FONTFACE.lines, OPTIONS); // get the length of this character

          lineLength += CharLength(FONTFACE.letterspace, FONTFACE.lines, OPTIONS) * OPTIONS.letterSpacing; // add letter spacing at the end
        }

        maxChars = 0; // new line, new maxLength goal

        output = AddLine(output, FONTFACE.lines, FONTFACE.buffer, OPTIONS.lineHeight); // adding new line
        // add letter spacing to the beginning

        output = AddLetterSpacing(output, FONTFACE.lines, FONTFACE.letterspace, FONTFACE.colors, OPTIONS.colors, OPTIONS.letterSpacing);
      }

      Debugging.report("lineLength at: \"".concat(lineLength, "\""), 2);

      if (CHAR !== "|") {
        maxChars++; // counting all printed characters

        output = AddChar(CHAR, output, FONTFACE.lines, FONTFACE.chars, FONTFACE.colors, OPTIONS.colors); // add new character

        output = AddLetterSpacing(output, FONTFACE.lines, FONTFACE.letterspace, FONTFACE.colors, OPTIONS.colors, OPTIONS.letterSpacing);
      }
    }

    output = AlignText(output, lineLength, FONTFACE.lines, OPTIONS.align, size); // alignment last line
  }

  var write = output.join("\n");

  if (FONTFACE.colors <= 1) {
    write = Colorize(write, FONTFACE.colors, OPTIONS.colors);
  }

  if (OPTIONS.space) {
    // add space
    write = "\n\n".concat(write, "\n\n");
  }

  if (OPTIONS.background !== 'transparent') {
    var _Color = Color(OPTIONS.background, true),
        openNew = _Color.open,
        closeNew = _Color.close;

    write = openNew + '\n' + write + closeNew; // result in one string with background
  }

  return {
    string: write,
    array: output,
    lines: lines,
    options: OPTIONS
  };
};

module.exports = exports = {
  Render: Render
};

File: rust/src/render.rs
//! The contents of this module is all about composing all functions together to render our output
extern crate enable_ansi_support;
use enable_ansi_support::enable_ansi_support;
use terminal_size::{terminal_size, Width};

use crate::chars::{
	add_letter, add_line, add_line_height, align_last_line, get_letter_length, get_letter_space, paint_letter,
};
use crate::color::{bgcolor2hex, get_background_color};
use crate::config::{Align, BgColors, Env, Fonts, Options};
use crate::debug::{d, Dt};
use crate::font;
use crate::gradient::add_gradient_colors;

/// The return struct you get from [`render()`]
pub struct RenderedString {
	/// A string ready to go to `stdout` with ansi escape codes and line breaks
	pub text: String,
	/// A vec of the string where each item is a new line in the output
	pub vec: Vec<String>,
	/// How many lines the output has
	/// – that is lines of cfonts text, not lines in output text
	pub lines: usize,
	/// The [`Options`] struct that was used to compile this output
	pub options: Options,
}

/// ## Render your text to a sexy font ready to be printed to `stdout`
///
/// Use this function if you want control over what to do with the output and
/// not have it automatically printed to `stdout` via the [`crate::say()`] function.
///
/// ```rust
/// extern crate cfonts;
///
/// use cfonts::{ render, Options, Fonts };
///
/// fn main() {
///     let output = render(Options {
///         text: String::from("hello"),
///         font: Fonts::FontTiny,
///         ..Options::default()
///     });
///
///     assert_eq!(
///         output.text,
///         format!("{}{}{}",
///             "\n\n",
///             " █ █ █▀▀ █   █   █▀█\n",
///             " █▀█ ██▄ █▄▄ █▄▄ █▄█\n\n"
///         )
///     );
/// }
/// ```
///
/// The return struct [`RenderedString`] will give you a bunch of things you
/// use to work with the output.
pub fn render(options: Options) -> RenderedString {
	d("render()", 1, Dt::Head, &options, &mut std::io::stdout());
	d(&format!("render() Options\n{:#?}", options), 1, Dt::Log, &options, &mut std::io::stdout());

	// enable ansi support in windows 10
	if let Ok(()) = enable_ansi_support() {}

	let size = terminal_size();
	let terminal_width = match options.env {
		Env::Cli => {
			if let Some((Width(w), _)) = size {
				w
			} else {
				80
			}
		}
		Env::Browser => 0xFFFF,
	};
	d(&format!("render()\nterminal_width:{:?}", terminal_width), 1, Dt::Log, &options, &mut std::io::stdout());

	let mut font = font::get(&options);
	let mut line_length = 0;
	let mut letter_count = 0;
	let mut lines = 0;
	let mut output: Vec<String> = Vec::new();

	// some fonts have smaller letter spacing
	let letter_spacing = if font.letterspace_size == 0 && options.letter_spacing > 0 {
		options.letter_spacing - 1
	} else {
		options.letter_spacing
	};
	d(&format!("render()\nletter_spacing:{:?}", letter_spacing), 1, Dt::Log, &options, &mut std::io::stdout());

	// the console font is special in that it has less line height space
	let line_height = if options.font == Fonts::FontConsole && options.line_height > 0 {
		options.line_height - 1
	} else {
		options.line_height
	};
	d(&format!("render()\nline_height:{:?}", line_height), 1, Dt::Log, &options, &mut std::io::stdout());

	let letter_space = get_letter_space(&font.letterspace, letter_spacing, &options);
	let letter_space_len = get_letter_length(&letter_space, font.colors, &options);
	let painted_letter_space = paint_letter(&letter_space, font.colors, &options);
	d(
		&format!(
			"render()\nletter_space:{:?}\nletter_space_len:{:?}\npainted_letter_space:{:?}",
			letter_space, letter_space_len, painted_letter_space
		),
		2,
		Dt::Log,
		&options,
		&mut std::io::stdout(),
	);

	add_line(&mut output, font.lines, &options);
	lines += 1;

	add_letter(&mut output, &font.buffer, &options);
	let buffer_len = get_letter_length(&font.buffer, font.lines, &options);
	line_length += buffer_len;
	d("render() added buffer", 1, Dt::Log, &options, &mut std::io::stdout());

	options.text.chars().for_each(|og_letter| {
		d(&format!("render() • loop og_letter:{:?}", og_letter), 1, Dt::Log, &options, &mut std::io::stdout());
		// we insert the pipe here so that out match will find it. it's not defined in our font files because it's just a new-line
		font.chars.insert(String::from("|"), vec![String::from("|")]);
		match font.chars.get(&og_letter.to_string().to_uppercase()) {
			None => {
				/* we ignore characters that are not supported */
				d(&format!("render() ignoring unknown letter:{:?}", og_letter), 1, Dt::Log, &options, &mut std::io::stdout());
			}
			Some(font_letter) => {
				d(&format!("render()\nfont_letter:{:?}", font_letter), 2, Dt::Log, &options, &mut std::io::stdout());
				let this_letter_len = get_letter_length(font_letter, font.colors, &options);
				if font_letter[0] == "|"
					|| this_letter_len + letter_space_len + line_length > terminal_width.into()
					|| letter_count + 1 > options.max_length && options.max_length > 0
				{
					d(
						&format!("render() added new line because\nletter == |:{:?}\nnew_length({:?}) > terminal_width({:?})\nnew_letter_count({:?}) > max_length({:?})",
							font_letter[0] == "|",
							this_letter_len + letter_space_len + line_length,
							terminal_width,
							letter_count + 1,
							options.max_length
						),
						1,
						Dt::Log,
						&options,
						&mut std::io::stdout()
					);
					align_last_line(&mut output, font.lines, line_length, terminal_width.into(), &options);
					d("render() aligned last line", 1, Dt::Log, &options, &mut std::io::stdout());
					add_line(&mut output, font.lines, &options);
					d("render() added new line", 1, Dt::Log, &options, &mut std::io::stdout());
					add_line_height(&mut output, line_height, &options);
					d("render() added line_height", 1, Dt::Log, &options, &mut std::io::stdout());
					add_letter(&mut output, &font.buffer, &options);
					d("render() added buffer", 1, Dt::Log, &options, &mut std::io::stdout());
					line_length = buffer_len;
					lines += 1;
					letter_count = 0;
				}

				if font_letter[0] != "|" {
					let painted_letter = paint_letter(font_letter, font.colors, &options);
					d("render() added color to letter", 1, Dt::Log, &options, &mut std::io::stdout());
					d(&format!("render()\npainted_letter:{:?}", painted_letter), 2, Dt::Log, &options, &mut std::io::stdout());
					add_letter(&mut output, &painted_letter_space, &options);
					d("render() added letter_space", 1, Dt::Log, &options, &mut std::io::stdout());
					add_letter(&mut output, &painted_letter, &options);
					d("render() added letter", 1, Dt::Log, &options, &mut std::io::stdout());
					letter_count += 1;
					line_length += letter_space_len + this_letter_len;
				}
			}
		}
	});
	align_last_line(&mut output, font.lines, line_length, terminal_width.into(), &options);
	d("render() aligned last line", 1, Dt::Log, &options, &mut std::io::stdout());

	if !options.gradient.is_empty() {
		output = add_gradient_colors(&output, lines, font.lines, &options);
		d("render() added gradient colors", 1, Dt::Log, &options, &mut std::io::stdout());
	}

	if !options.spaceless {
		match options.align {
			Align::Top => output.push(String::from("\n\n\n")),
			Align::Bottom => output[0] = String::from("\n\n\n\n") + &output[0],
			_ => {
				output[0] = String::from("\n\n") + &output[0];
				output.push(String::from("\n"));
			}
		}
		d("render() added space", 1, Dt::Log, &options, &mut std::io::stdout());
	}

	if options.background != BgColors::Transparent && options.env == Env::Cli {
		let (open, close) = get_background_color(&options.background);
		output[0] = format!("{}\n", open) + &output[0];
		let last_index = output.len() - 1;
		output[last_index] = format!("{}{}", output[output.len() - 1], close);
		d("render() added background", 1, Dt::Log, &options, &mut std::io::stdout());
	}

	let mut text = match options.env {
		Env::Cli => output.join("\n"),
		Env::Browser => output.join("<br>"),
	};

	if options.env == Env::Browser {
		let color = bgcolor2hex(&options.background, &options);
		let align = match options.align {
			Align::Left => "left",
			Align::Right => "right",
			Align::Center => "center",
			_ => "left",
		};
		text = format!("<div style=\"font-family:monospace;white-space:pre;text-align:{};max-width:100%;overflow:scroll;background:{}\">{}<div>", align, color, text);
		d("render() formatted for Env::Browser", 1, Dt::Log, &options, &mut std::io::stdout());
	}

	RenderedString {
		text,
		vec: output,
		lines,
		options,
	}
}
